pkg/server/settingsworker_test.go

```diff
@@ -375,3 +379,111 @@ func TestSettingsZipkinBatchSize(t *testing.T) {
 	}
 	assert.Equal(t, observedBatchReport.ObservedBatchSize, 1)
 }
+
+// TestSettingsTraceSamplingRate tests that setting the sampling rate
+// of traces will correctly limit the percentage of traces sampled.
+func TestSettingsTraceSamplingRate(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// Start test cluster containing a test db and test table
+	ctx := context.Background()
+	tc := testcluster.StartTestCluster(t, 3, base.TestClusterArgs{})
+	defer tc.Stopper().Stop(ctx)
+
+	execStatement := func(statement string) {
+		if _, err := tc.ServerConn(rand.Intn(3)).Exec(statement); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	execStatement(`CREATE DATABASE t`)
+	execStatement(`CREATE TABLE test (k INT PRIMARY KEY, v INT)`)
+
+	// Set up tracing collector and variable to count the number of traces
+	// collected from INSERT INTO statements
+	nTraces := int64(0)
+	collectorHandlerFcn := func(rw http.ResponseWriter, req *http.Request) {
+		bodyBytes, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			return
+		}
+		t := thrift.NewTMemoryBuffer()
+		if _, err = t.Write(bodyBytes); err != nil {
+			return
+		}
+		p := thrift.NewTBinaryProtocolTransport(t)
+		_, batchSize, err := p.ReadListBegin()
+		if err != nil {
+			return
+		}
+
+		for i := 0; i < batchSize; i += 1 {
+			span := zipkincore.Span{}
+			if err = span.Read(p); err != nil {
+				return
+			}
+
+			if span.Name == `exec stmt` {
+				for _, annotation := range span.Annotations {
+					if strings.Contains(annotation.Value, `INSERT INTO test VALUES`) &&
+						strings.Contains(annotation.Value, `in state: Open`) {
+						atomic.AddInt64(&nTraces, 1)
+					}
+				}
+			}
+		}
+	}
+	traceCollectorServer := httptest.NewServer(http.HandlerFunc(collectorHandlerFcn))
+	defer traceCollectorServer.Close()
+
+	// Configure tracing in the test cluster
+	traceCollectorServerURL, err := url.Parse(traceCollectorServer.URL)
+	if err != nil {
+		t.Fatal(err)
+	}
+	setClusterTraceSamplingRate := func(rate float64) {
+		execStatement(fmt.Sprintf(`SET CLUSTER SETTING trace.zipkin.sampling.rate = %f`, rate))
+	}
+	execStatement(fmt.Sprintf(`SET CLUSTER SETTING trace.zipkin.collector = '%s'`, traceCollectorServerURL.Host))
```

Comment:
The test uses the default batch size of 100. Setting it to 0 actually caused problems because spans would get dropped when many were being created in parallel (apparently, the collector uses a buffer which has the same length as the batch size we set).

Although the sampling rate limiting feature did not use configurable batch sizes, I figured that the batch size setting does not hurt, and we might as well keep it as it might be useful in the future.